src/main/java/org/jenkinsci/test/acceptance/docker/DockerImage.java

```diff
@@ -92,45 +89,56 @@ public DockerImage(String tag) {
         docker.add(starter.args);
 
         File logfile = starter.log;
-        if (logfile == null) {
-            logfile = new File(cidFile + ".log");
-        }
 
-        System.out.printf("Launching Docker container `%s`: logfile is at %s\n", docker.toString(), logfile);
+        System.out.printf("Launching Docker container `%s`: logfile will be at %s\n", docker.toString(), logfile);
 
         Process p = docker.build()
                 .redirectInput(new File(SystemUtils.IS_OS_WINDOWS ? "NUL": "/dev/null"))
                 .redirectErrorStream(true)
-                .redirectOutput(logfile)
                 .start();
 
-        String cid = waitForCid(docker, cidFile, logfile, p);
+
+        String cid = waitForCid(docker, p);
+
+        Process logProcess = Docker.cmd("logs")
+                .add("-f")
+                .add(cid)
+                .build()
+                .redirectInput(new File(SystemUtils.IS_OS_WINDOWS ? "NUL": "/dev/null"))
+                .redirectErrorStream(true)
+                .redirectOutput(logfile)
+                .start();
 
         try {
             T t = type.newInstance();
-            t.init(cid, p, logfile);
+            t.init(cid, logProcess, logfile);
             return t;
         } catch (ReflectiveOperationException e) {
             throw new Error(e);
         }
     }
 
-    private String waitForCid(CommandBuilder docker, File cidFile, File logfile, Process p) throws InterruptedException, IOException {
+    private String waitForCid(CommandBuilder docker, Process p) throws InterruptedException, IOException {
         for (int i = 0; i < 10; i++) {
-            Thread.sleep(1000);
+            Thread.sleep(500);
 
-            String cid = FileUtils.readFileToString(cidFile);
-            if (cid != null && cid.length() != 0) return cid;
+            String output = IOUtils.toString(p.getInputStream());
 
             try {
-                p.exitValue();
-                throw new IOException("docker died unexpectedly: "+docker+"\n"+ FileUtils.readFileToString(logfile));
+                if (p.exitValue() != 0) {
+                    throw new IOException("docker died unexpectedly: " + docker + "\n" + output);
+                }
             } catch (IllegalThreadStateException e) {
-                //Docker is still running okay.
+                // Docker is still launching, so we move on.
```

Comment:
What - why...?
THis should not happen as
1) you have read the entire stream to a String (it should not return until you get EOF?)
2) if this could happen use `waitFor()`